sycl/include/sycl/ext/oneapi/matrix/matrix-jit-use.hpp

```diff
@@ -0,0 +1,517 @@
+          access::address_space Space>
+inline __SYCL_ALWAYS_INLINE void
+joint_matrix_store(Group sg,
+                   joint_matrix<T, NumRows, NumCols, Use, MatL, Group> &src,
+                   multi_ptr<T, Space> res, size_t stride, matrix_layout MemL) {
+#ifdef __SYCL_DEVICE_ONLY__
+  T *Ptr = res.get();
+  switch (MemL) {
+  default:
+    assert(false && "Invalid Memory Layout!");
+  case matrix_layout::row_major:
+    __spirv_JointMatrixStoreINTEL<T, NumRows, NumCols,
+                                  spv_matrix_use_traits<Use>::value,
+                                  spv_matrix_layout_traits<MatL>::value>(
+        Ptr, src.spvm, stride, __spv::MatrixLayout::RowMajor,
+        spv_scope_traits<Group>::value);
+    break;
+  case matrix_layout::col_major:
+    __spirv_JointMatrixStoreINTEL<T, NumRows, NumCols,
+                                  spv_matrix_use_traits<Use>::value,
+                                  spv_matrix_layout_traits<MatL>::value>(
+        Ptr, src.spvm, stride, __spv::MatrixLayout::ColumnMajor,
+        spv_scope_traits<Group>::value);
+    break;
+  case matrix_layout::packed_a:
+    __spirv_JointMatrixStoreINTEL<T, NumRows, NumCols,
+                                  spv_matrix_use_traits<Use>::value,
+                                  spv_matrix_layout_traits<MatL>::value>(
+        Ptr, src.spvm, stride, __spv::MatrixLayout::PackedA,
+        spv_scope_traits<Group>::value);
+    break;
+  case matrix_layout::packed_b:
+    __spirv_JointMatrixStoreINTEL<T, NumRows, NumCols,
+                                  spv_matrix_use_traits<Use>::value,
+                                  spv_matrix_layout_traits<MatL>::value>(
+        Ptr, src.spvm, stride, __spv::MatrixLayout::PackedB,
+        spv_scope_traits<Group>::value);
+    break;
+  }
+#else
+  (void)sg;
+  (void)src;
+  (void)res;
+  (void)stride;
+  (void)MemL;
+  throw runtime_error("joint matrix is not supported on host device.",
+                      PI_INVALID_DEVICE);
+#endif // __SYCL_DEVICE_ONLY__
+}
+
+template <typename Group, typename T1, typename T2, typename T3, size_t M,
+          size_t K, size_t N, matrix_use UseA, matrix_use UseB, matrix_use UseC,
+          matrix_layout LayoutA, matrix_layout LayoutB, matrix_layout LayoutC>
+inline __SYCL_ALWAYS_INLINE joint_matrix<T3, M, N, UseC, LayoutC, Group>
+joint_matrix_mad(Group sg, joint_matrix<T1, M, K, UseA, LayoutA, Group> &mA,
+                 joint_matrix<T2, K, N, UseB, LayoutB, Group> &mB,
+                 joint_matrix<T3, M, N, UseC, LayoutC, Group> &mC) {
+#ifdef __SYCL_DEVICE_ONLY__
+  joint_matrix<T3, M, N, UseC, LayoutC, Group> res(sg);
+  if constexpr (std::is_same<T1, uint16_t>::value &&
+                std::is_same<T2, uint16_t>::value &&
+                std::is_same<T3, float>::value)
+    res.spvm = __spirv_JointMatrixMadINTEL(mA.spvm, mB.spvm, mC.spvm);
+  else if constexpr (std::is_unsigned<T1>::value && std::is_unsigned<T2>::value)
```

Comment:
```suggestion
  else if constexpr (std::is_unsigned<T1>::value && std::is_unsigned<T2>::value)
```
```suggestion
  else if constexpr (std::is_unsigned_v<T1> && std::is_unsigned_v<T2>)
```

Superficial implementation detail:
Switching to c++17: note that this suggested change leads to basic_tests/stdcpp_compat.cpp to fail but there is a procedure mentioned somewhere to prevent this fail: I believe we add a flag somewhere. Apparently we should support c++17 everywhere however: https://github.com/intel/llvm/pull/5964#discussion_r848913703